app/src/main/java/co/tinode/tindroid/db/MessageDb.java

```diff
@@ -218,13 +221,49 @@ public static Cursor query(SQLiteDatabase db, long topicId, int pageCount, int p
                     " AND "
                         + COLUMN_NAME_STATUS + "<=" + BaseDb.STATUS_VISIBLE +
                 " ORDER BY "
-                    + COLUMN_NAME_TS + " DESC, "
                     + COLUMN_NAME_SEQ + " DESC" +
                 " LIMIT " + (pageCount * pageSize);
 
         return db.rawQuery(sql, null);
     }
 
+    /**
+     * Query for the latest messages. The query returns NULL-filled rows whenever there is a gap in seq values.
+     * See explanation here: https://stackoverflow.com/questions/31589843/how-do-i-find-gap-in-sqlite-table
+     *
+     * @param db        database to select from;
+     * @param topicId   Tinode topic ID (topics._id) to select from
+     * @param maxSeq    Maximum known seqId in the table.
+     * @param pageCount number of pages to return
+     * @param pageSize  number of messages per page
+     * @return cursor with the messages.
+     */
+    public static Cursor query(SQLiteDatabase db, long topicId, int maxSeq, int pageCount, int pageSize) {
+        final String sql ="SELECT " +
+                    _ID + "," + COLUMN_NAME_TOPIC_ID + "," + COLUMN_NAME_USER_ID + "," + COLUMN_NAME_STATUS + "," +
+                    COLUMN_NAME_SENDER + "," + COLUMN_NAME_TS + "," + COLUMN_NAME_SEQ + "," + COLUMN_NAME_HEAD + "," +
+                    COLUMN_NAME_CONTENT +
+                " FROM " + TABLE_NAME +
+                " WHERE " + COLUMN_NAME_TOPIC_ID + "=" + topicId +
+                " AND " + COLUMN_NAME_STATUS + "<=" + BaseDb.STATUS_VISIBLE +
+                " UNION ALL " +
+                // The following query returns gaps.
```

Comment:
I understand it is a design decision but what were the factors you took into account when choosing this approach?
Another approach could be adding a deletion flag field in the Messages table and inserting empty rows into it in place of deleted rows. The read logic (which is executed more frequently than deletions) thus becomes more simple.

Reply:
The main question is UX. I think it's desirable to
1. Show to the user when messages are deleted
2. When multiple sequential messages are deleted, show just one indicator, not an indicator for each deleted message.

> adding a deletion flag field in the Messages table 

That was the original approach. It has two drawbacks: it does not satisfy (2); when messages are deleted before the user joins the conversation, the server never sends deleted messages to the client. The client must infer that a bunch of messages were deleted and insert empty placeholders.

> and inserting empty rows into it in place of deleted rows

That was the second approach. It also does not satisfy 2.